File: python/src/chiptime/metrics/insights.py
"""Single-workout report + machine-readable insights.

Insights follow the error-code philosophy (contract #5): a stable CODE for
agents, a human sentence, and numeric evidence. An insight only appears
when the data actually shows it; raw numbers live on the report itself.
Analyses that need absent inputs land in `omissions` with a reason —
a missing number beats an invented one (ADR-0008 §4/§5).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from chiptime.message import Message
from chiptime.metrics._basics import mean_max, swolf, time_in_zones
from chiptime.metrics.intervals import IntervalStructure, detect_structure
from chiptime.metrics.load import (
    TRIMP_MIN_COVERAGE,
    LoadEstimate,
    hr_coverage_fraction,
    weighted_avg_power,
    work_kj,
    workout_load,
)
from chiptime.metrics.pacing import Split, distance_splits, format_pace, session_pace_s
from chiptime.metrics.settings import AthleteSettings
from chiptime.metrics.sports import cadence_display, primary_signal, profile_for
from chiptime.metrics.zones import hr_zone_bounds, power_zone_bounds
from chiptime.model import Session

# --- insight thresholds (TS port copies verbatim) ------------------------
SPLIT_DELTA_PCT = 2.0  # first-vs-second-half speed delta to call a split
HR_DRIFT_PCT = 5.0  # efficiency drop that suggests aerobic fatigue
COASTING_SHARE_PCT = 25.0  # zero-power share worth remarking on (rides)
MIN_HALF_SAMPLES = 60  # halves need this many paired samples to compare
POWER_CURVE_WINDOWS = [5, 60, 300, 1200]

INSIGHT_CODES: dict[str, str] = {
    "PACING_NEGATIVE_SPLIT": "Second half faster than the first by more than 2%",
    "PACING_POSITIVE_SPLIT": "Second half slower than the first by more than 2%",
    "HR_DRIFT_HIGH": "Speed/power per heartbeat fell >5% first half to second (aerobic decoupling)",
    "COASTING_HIGH": "More than 25% of ride samples at 0 W",
    "WORKOUT_STRUCTURE": "Repeated interval structure found (label in evidence)",
}


@dataclass(frozen=True, slots=True)
class Insight:
    """One notable observation: a stable machine ``code`` (see
    `INSIGHT_CODES`), a human sentence, and the numbers behind it."""

    code: str
    message: str
    evidence: dict[str, Any]


@dataclass(slots=True)
class WorkoutReport:
    """Everything is optional and null-honest; `omissions` says what was not
    computed and why. `basis` strings mark where derived numbers came from."""

    sport: str
    sub_sport: str | None
    profile: str
    primary_signal: str  # power | speed | none
    duration_s: dict[str, float | None] = field(default_factory=dict)
    distance_m: float | None = None
    pace: dict[str, Any] | None = None  # {seconds, style, formatted, basis}
    avg_speed_kmh: float | None = None
    avg_speed_basis: str | None = None
    avg_primary: float | None = None
    max_primary: float | None = None
    avg_hr: float | None = None
    max_hr: float | None = None
    cadence: dict[str, Any] | None = None  # {value, units, note}
    weighted_avg_power: float | None = None
    variability_ratio: float | None = None  # weighted / avg power
    work_kj: float | None = None
    power_curve: dict[int, float | None] | None = None
    swolf: float | None = None
    splits: list[Split] = field(default_factory=list)
    structure: IntervalStructure | None = None
    hr_zones: dict[str, Any] | None = None  # {bounds, seconds, basis}
    power_zones: dict[str, Any] | None = None
    load: LoadEstimate | None = None
    insights: list[Insight] = field(default_factory=list)
    omissions: list[str] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        return _plain(self)  # type: ignore[no-any-return]


@dataclass(slots=True)
class ActivityReport:
    """The full-file report: one `WorkoutReport` per session, in order.
    ``to_dict()`` gives the JSON-ready form the CLI emits with ``--json``."""

    sessions: list[WorkoutReport]

    def to_dict(self) -> dict[str, Any]:
        return {"sessions": [s.to_dict() for s in self.sessions]}


def _plain(obj: Any) -> Any:
    """Dataclass -> JSON-ready plain data (deterministic key order comes from
    dataclass field order + sort at dump time)."""
    import dataclasses
    from datetime import date, datetime

    if dataclasses.is_dataclass(obj) and not isinstance(obj, type):
        return {f.name: _plain(getattr(obj, f.name)) for f in dataclasses.fields(obj)}
    if isinstance(obj, dict):
        return {str(k): _plain(v) for k, v in obj.items()}
    if isinstance(obj, (list, tuple)):
        return [_plain(v) for v in obj]
    if isinstance(obj, (datetime, date)):
        return obj.isoformat()
    return obj


# --- half comparisons ----------------------------------------------------


def _halves(session: Session, stream_name: str) -> tuple[list[float], list[float]]:
    s = session.records.stream(stream_name)
    if s is None:
        return [], []
    vals = [float(v) for v in s.values if isinstance(v, (int, float)) and not isinstance(v, bool)]
    mid = len(vals) // 2
    return vals[:mid], vals[mid:]


def _pacing_insight(session: Session, stream: str | None) -> Insight | None:
    if stream is None:
        return None
    first, second = _halves(session, stream)
    if len(first) < MIN_HALF_SAMPLES or len(second) < MIN_HALF_SAMPLES:
        return None
    a = sum(first) / len(first)
    b = sum(second) / len(second)
    if a <= 0:
        return None
    delta_pct = (b - a) / a * 100.0
    ev = {
        "first_half_avg": round(a, 3),
        "second_half_avg": round(b, 3),
        "delta_pct": round(delta_pct, 1),
        "stream": stream,
    }
    if delta_pct >= SPLIT_DELTA_PCT:
        return Insight(
            "PACING_NEGATIVE_SPLIT", f"Second half {delta_pct:.1f}% faster than the first.", ev
        )
    if delta_pct <= -SPLIT_DELTA_PCT:
        return Insight(
            "PACING_POSITIVE_SPLIT", f"Second half {abs(delta_pct):.1f}% slower than the first.", ev
        )
    return None


def _hr_drift_insight(session: Session, stream: str | None) -> Insight | None:
    if stream is None:
        return None
    eff = session.records.stream(stream)
    hr = session.records.stream("heart_rate")
    if eff is None or hr is None:
        return None
    pairs = [
        (float(e), float(h))
        for e, h in zip(eff.values, hr.values, strict=True)
        if isinstance(e, (int, float))
        and not isinstance(e, bool)
        and isinstance(h, (int, float))
        and not isinstance(h, bool)
        and h > 0
    ]
    mid = len(pairs) // 2
    if mid < MIN_HALF_SAMPLES:
        return None

    def ef(chunk: list[tuple[float, float]]) -> float:
        se = sum(e for e, _ in chunk)
        sh = sum(h for _, h in chunk)
        return se / sh if sh > 0 else 0.0

    ef1, ef2 = ef(pairs[:mid]), ef(pairs[mid:])
    if ef1 <= 0:
        return None
    drift_pct = (ef1 - ef2) / ef1 * 100.0
    if drift_pct <= HR_DRIFT_PCT:
        return None
    return Insight(
        "HR_DRIFT_HIGH",
        f"Output per heartbeat fell {drift_pct:.1f}% from first half to second — aerobic drift.",
        {
            "drift_pct": round(drift_pct, 1),
            "stream": stream,
            "first_half_ef": round(ef1, 4),
            "second_half_ef": round(ef2, 4),
        },
    )


def _coasting_insight(session: Session) -> Insight | None:
    if profile_for(session).key != "cycling":
        return None
    pw = session.records.stream("power")
    if pw is None:
        return None
    present = [
        float(v) for v in pw.values if isinstance(v, (int, float)) and not isinstance(v, bool)
    ]
    if len(present) < MIN_HALF_SAMPLES:
        return None
    zero_share = sum(1 for v in present if v == 0.0) / len(present) * 100.0
    if zero_share < COASTING_SHARE_PCT:
        return None
    return Insight(
        "COASTING_HIGH",
        f"{zero_share:.0f}% of samples at 0 W (coasting).",
        {"zero_share_pct": round(zero_share, 1)},
    )


# --- report builder ------------------------------------------------------


def analyze_session(
    session: Session, messages: list[Message] | None = None, settings: AthleteSettings | None = None
) -> WorkoutReport:
    profile = profile_for(session)
    kind, stream = primary_signal(session)
    rep = WorkoutReport(
        sport=session.sport, sub_sport=session.sub_sport, profile=profile.key, primary_signal=kind
    )

    der, dec = session.derived, session.declared
    rep.duration_s = {
        "elapsed": der.elapsed_time_s or (dec.elapsed_time_s if dec else None),
        "timer": der.timer_time_s or (dec.timer_time_s if dec else None),
        "moving": der.moving_time_s,
    }
    rep.distance_m = der.distance_m or (dec.distance_m if dec else None)

    # pace / speed presentation per profile
    got = session_pace_s(session, profile.pace_style) if profile.pace_style != "speed" else None
    if got is not None:
        pace_s, basis = got
        rep.pace = {
            "seconds": round(pace_s, 1),
            "style": profile.pace_style,
            "formatted": format_pace(pace_s, profile.pace_style, suffix=True),
            "basis": basis,
        }
    if rep.distance_m:
        for key in ("moving", "timer", "elapsed"):
            dur_for_speed = rep.duration_s.get(key)
            if dur_for_speed:
                rep.avg_speed_kmh = round(rep.distance_m / dur_for_speed * 3.6, 2)
                rep.avg_speed_basis = key
                break

    # primary + hr stats from streams (avg dicts as declared fallback)
    def _stream_stats(name: str) -> tuple[float | None, float | None]:
        s = session.records.stream(name)
        if s is None:
            return None, None
        vals = [
            float(v) for v in s.values if isinstance(v, (int, float)) and not isinstance(v, bool)
        ]
        if not vals:
            return None, None
        return sum(vals) / len(vals), max(vals)

    if stream is not None:
        rep.avg_primary, rep.max_primary = _stream_stats(stream)
    rep.avg_hr, rep.max_hr = _stream_stats("heart_rate")

    cad_avg = _stream_stats("cadence")[0]
    if cad_avg is not None:
        val, units, note = cadence_display(cad_avg, profile)
        rep.cadence = {
            "value": round(val, 1) if val is not None else None,
            "units": units,
            "note": note,
        }

    pw = session.records.stream("power")
    if pw is not None:
        rep.weighted_avg_power = weighted_avg_power(pw.values)
        if rep.weighted_avg_power is not None:
            rep.weighted_avg_power = round(rep.weighted_avg_power, 1)
        if rep.weighted_avg_power and rep.avg_primary and kind == "power":
            rep.variability_ratio = round(rep.weighted_avg_power / rep.avg_primary, 3)
        kj = work_kj(session.records.time, pw.values)
        rep.work_kj = round(kj, 1) if kj is not None else None
        curve = mean_max(pw.values, POWER_CURVE_WINDOWS)
        rep.power_curve = {w: (round(v, 1) if v is not None else None) for w, v in curve.items()}

    if profile.distance_from_lengths and session.lengths:
        rep.swolf = swolf(session)[1]

    rep.splits = (
        distance_splits(session, 1000.0, style=profile.pace_style)
        if profile.pace_style == "per_km"
        else []
    )
    rep.structure = detect_structure(session, messages, settings)

    # zones: only from settings or the file, never estimated (ADR-0008 §4)
    hb, hbasis = hr_zone_bounds(settings, messages)
    if hb is not None and session.records.stream("heart_rate") is not None:
        hr_stream = session.records.stream("heart_rate")
        assert hr_stream is not None
        rep.hr_zones = {
            "bounds": list(hb),
            "basis": hbasis,
            "seconds": time_in_zones(session.records.time, hr_stream.values, list(hb)),
        }
    elif session.records.stream("heart_rate") is not None:
        rep.omissions.append("hr_zones: no zone bounds in settings or file")
    pb, pbasis = power_zone_bounds(settings, messages)
    if pb is not None and pw is not None:
        rep.power_zones = {
            "bounds": list(pb),
            "basis": pbasis,
            "seconds": time_in_zones(session.records.time, pw.values, list(pb)),
        }
    elif pw is not None:
        rep.omissions.append("power_zones: no zone bounds in settings or file")

    rep.load = workout_load(session, settings)
    if rep.load is None:
        if pw is not None and (settings is None or not settings.ftp_w):
            rep.omissions.append("load: power present but no ftp_w in settings")
        elif session.records.stream("heart_rate") is not None:
            if settings is None or not (settings.max_hr and settings.resting_hr):
                rep.omissions.append("load: hr present but no max_hr+resting_hr in settings")
            else:
                cov = hr_coverage_fraction(session)
                if cov is not None and cov < TRIMP_MIN_COVERAGE:
                    rep.omissions.append(
                        f"load: hr covers only {cov:.0%} of the session "
                        f"(< {TRIMP_MIN_COVERAGE:.0%}); trimp would understate load"
                    )

    for maybe in (
        _pacing_insight(session, stream),
        _hr_drift_insight(session, stream),
        _coasting_insight(session),
    ):
        if maybe is not None:
            rep.insights.append(maybe)
    if rep.structure is not None and rep.structure.repeats:
        rep.insights.append(
            Insight(
                "WORKOUT_STRUCTURE",
                "Interval structure: " + "; ".join(g.label for g in rep.structure.repeats),
                {"basis": rep.structure.basis, "labels": [g.label for g in rep.structure.repeats]},
            )
        )
    return rep


def analyze(result: Any, settings: AthleteSettings | None = None) -> ActivityReport:
    """Report per session from a ParseResult (uses .activity and .messages)."""
    activity = result.activity
    messages = result.messages
    if activity is None:
        return ActivityReport(sessions=[])
    return ActivityReport(
        sessions=[analyze_session(s, messages, settings) for s in activity.sessions]
    )

File: js/src/metrics/insights.ts
/**
 * Single-workout report + machine-readable insights.
 *
 * Twin of `python/src/chiptime/metrics/insights.py`. Insights follow the error-code
 * philosophy (contract #5): a stable CODE for agents, a human sentence, and numeric
 * evidence. Analyses that need absent inputs land in `omissions` with a reason.
 */

import { fitTsToIso } from "../decode.js";
import type { Message } from "../message.js";
import { type Session, getStream } from "../model.js";
import { pyFixed, pyRound, pyRoundN, pySum } from "../numeric.js";
import { meanMax, swolf, timeInZones } from "./basics.js";
import { type IntervalStructure, detectStructure } from "./intervals.js";
import {
  type LoadEstimate,
  TRIMP_MIN_COVERAGE,
  hrCoverageFraction,
  weightedAvgPower,
  workKj,
  workoutLoad,
} from "./load.js";
import { type Split, distanceSplits, formatPace, sessionPaceS } from "./pacing.js";
import type { AthleteSettings } from "./settings.js";
import { cadenceDisplay, primarySignal, profileFor } from "./sports.js";
import { hrZoneBounds, powerZoneBounds } from "./zones.js";

// --- insight thresholds (copied verbatim) ---------------------------------
export const SPLIT_DELTA_PCT = 2.0;
export const HR_DRIFT_PCT = 5.0;
export const COASTING_SHARE_PCT = 25.0;
const MIN_HALF_SAMPLES = 60;
const POWER_CURVE_WINDOWS = [5, 60, 300, 1200];

export const INSIGHT_CODES: Readonly<Record<string, string>> = {
  PACING_NEGATIVE_SPLIT: "Second half faster than the first by more than 2%",
  PACING_POSITIVE_SPLIT: "Second half slower than the first by more than 2%",
  HR_DRIFT_HIGH: "Speed/power per heartbeat fell >5% first half to second (aerobic decoupling)",
  COASTING_HIGH: "More than 25% of ride samples at 0 W",
  WORKOUT_STRUCTURE: "Repeated interval structure found (label in evidence)",
};

export interface Insight {
  readonly code: string;
  readonly message: string;
  readonly evidence: Readonly<Record<string, unknown>>;
}

/** Null-honest; `omissions` says what was not computed and why. */
export interface WorkoutReport {
  sport: string;
  subSport: string | null;
  profile: string;
  primarySignal: string;
  durationS: Map<string, number | null>;
  distanceM: number | null;
  pace: Record<string, unknown> | null;
  avgSpeedKmh: number | null;
  avgSpeedBasis: string | null;
  avgPrimary: number | null;
  maxPrimary: number | null;
  avgHr: number | null;
  maxHr: number | null;
  cadence: Record<string, unknown> | null;
  weightedAvgPower: number | null;
  variabilityRatio: number | null;
  workKj: number | null;
  powerCurve: Map<number, number | null> | null;
  swolf: number | null;
  splits: Split[];
  structure: IntervalStructure | null;
  hrZones: Record<string, unknown> | null;
  powerZones: Record<string, unknown> | null;
  load: LoadEstimate | null;
  insights: Insight[];
  omissions: string[];
}

export interface ActivityReport {
  sessions: WorkoutReport[];
}

// --- half comparisons -----------------------------------------------------

function halves(session: Session, streamName: string): [number[], number[]] {
  const s = getStream(session.records, streamName);
  if (s === undefined) return [[], []];
  const vals = s.values.filter((v): v is number => typeof v === "number");
  const mid = Math.floor(vals.length / 2);
  return [vals.slice(0, mid), vals.slice(mid)];
}

function pacingInsight(session: Session, stream: string | null): Insight | null {
  if (stream === null) return null;
  const [first, second] = halves(session, stream);
  if (first.length < MIN_HALF_SAMPLES || second.length < MIN_HALF_SAMPLES) return null;
  const a = pySum(first) / first.length;
  const b = pySum(second) / second.length;
  if (a <= 0) return null;
  const deltaPct = ((b - a) / a) * 100.0;
  const ev = {
    first_half_avg: pyRoundN(a, 3),
    second_half_avg: pyRoundN(b, 3),
    delta_pct: pyRoundN(deltaPct, 1),
    stream,
  };
  if (deltaPct >= SPLIT_DELTA_PCT) {
    return {
      code: "PACING_NEGATIVE_SPLIT",
      message: `Second half ${pyFixed(deltaPct, 1)}% faster than the first.`,
      evidence: ev,
    };
  }
  if (deltaPct <= -SPLIT_DELTA_PCT) {
    return {
      code: "PACING_POSITIVE_SPLIT",
      message: `Second half ${pyFixed(Math.abs(deltaPct), 1)}% slower than the first.`,
      evidence: ev,
    };
  }
  return null;
}

function hrDriftInsight(session: Session, stream: string | null): Insight | null {
  if (stream === null) return null;
  const eff = getStream(session.records, stream);
  const hr = getStream(session.records, "heart_rate");
  if (eff === undefined || hr === undefined) return null;
  const pairs: [number, number][] = [];
  for (let i = 0; i < eff.values.length; i++) {
    const e = eff.values[i];
    const h = hr.values[i];
    if (typeof e === "number" && typeof h === "number" && h > 0) pairs.push([e, h]);
  }
  const mid = Math.floor(pairs.length / 2);
  if (mid < MIN_HALF_SAMPLES) return null;

  const ef = (chunk: [number, number][]): number => {
    const se = pySum(chunk.map(([e]) => e));
    const sh = pySum(chunk.map(([, h]) => h));
    return sh > 0 ? se / sh : 0;
  };

  const ef1 = ef(pairs.slice(0, mid));
  const ef2 = ef(pairs.slice(mid));
  if (ef1 <= 0) return null;
  const driftPct = ((ef1 - ef2) / ef1) * 100.0;
  if (driftPct <= HR_DRIFT_PCT) return null;
  return {
    code: "HR_DRIFT_HIGH",
    message: `Output per heartbeat fell ${pyFixed(driftPct, 1)}% from first half to second — aerobic drift.`,
    evidence: {
      drift_pct: pyRoundN(driftPct, 1),
      stream,
      first_half_ef: pyRoundN(ef1, 4),
      second_half_ef: pyRoundN(ef2, 4),
    },
  };
}

function coastingInsight(session: Session): Insight | null {
  if (profileFor(session).key !== "cycling") return null;
  const pw = getStream(session.records, "power");
  if (pw === undefined) return null;
  const present = pw.values.filter((v): v is number => typeof v === "number");
  if (present.length < MIN_HALF_SAMPLES) return null;
  let zeros = 0;
  for (const v of present) if (v === 0) zeros++;
  const zeroShare = (zeros / present.length) * 100.0;
  if (zeroShare < COASTING_SHARE_PCT) return null;
  return {
    code: "COASTING_HIGH",
    message: `${pyFixed(zeroShare, 0)}% of samples at 0 W (coasting).`,
    evidence: { zero_share_pct: pyRoundN(zeroShare, 1) },
  };
}

// --- report builder -------------------------------------------------------

export function analyzeSession(
  session: Session,
  messages: readonly Message[] | null = null,
  settings: AthleteSettings | null = null,
): WorkoutReport {
  const profile = profileFor(session);
  const [kind, stream] = primarySignal(session);
  const rep: WorkoutReport = {
    sport: session.sport,
    subSport: session.subSport,
    profile: profile.key,
    primarySignal: kind,
    durationS: new Map(),
    distanceM: null,
    pace: null,
    avgSpeedKmh: null,
    avgSpeedBasis: null,
    avgPrimary: null,
    maxPrimary: null,
    avgHr: null,
    maxHr: null,
    cadence: null,
    weightedAvgPower: null,
    variabilityRatio: null,
    workKj: null,
    powerCurve: null,
    swolf: null,
    splits: [],
    structure: null,
    hrZones: null,
    powerZones: null,
    load: null,
    insights: [],
    omissions: [],
  };

  const der = session.derived;
  const dec = session.declared;
  // Python `or`: 0.0 falls through to the declared value.
  rep.durationS.set("elapsed", der.elapsedTimeS || (dec ? dec.elapsedTimeS : null));
  rep.durationS.set("timer", der.timerTimeS || (dec ? dec.timerTimeS : null));
  rep.durationS.set("moving", der.movingTimeS);
  rep.distanceM = der.distanceM || (dec ? dec.distanceM : null);

  // pace / speed presentation per profile
  const got = profile.paceStyle !== "speed" ? sessionPaceS(session, profile.paceStyle) : null;
  if (got !== null) {
    const [paceS, basis] = got;
    rep.pace = {
      seconds: pyRoundN(paceS, 1),
      style: profile.paceStyle,
      formatted: formatPace(paceS, profile.paceStyle, { suffix: true }),
      basis,
    };
  }
  if (rep.distanceM) {
    for (const key of ["moving", "timer", "elapsed"]) {
      const durForSpeed = rep.durationS.get(key);
      if (durForSpeed) {
        rep.avgSpeedKmh = pyRoundN((rep.distanceM / durForSpeed) * 3.6, 2);
        rep.avgSpeedBasis = key;
        break;
      }
    }
  }

  const streamStats = (name: string): [number | null, number | null] => {
    const s = getStream(session.records, name);
    if (s === undefined) return [null, null];
    const vals = s.values.filter((v): v is number => typeof v === "number");
    if (vals.length === 0) return [null, null];
    let mx = vals[0] as number;
    for (const v of vals) if (v > mx) mx = v;
    return [pySum(vals) / vals.length, mx];
  };

  if (stream !== null) [rep.avgPrimary, rep.maxPrimary] = streamStats(stream);
  [rep.avgHr, rep.maxHr] = streamStats("heart_rate");

  const cadAvg = streamStats("cadence")[0];
  if (cadAvg !== null) {
    const [val, units, note] = cadenceDisplay(cadAvg, profile);
    rep.cadence = { value: val !== null ? pyRoundN(val, 1) : null, units, note };
  }

  const pw = getStream(session.records, "power");
  if (pw !== undefined) {
    rep.weightedAvgPower = weightedAvgPower(pw.values);
    if (rep.weightedAvgPower !== null) rep.weightedAvgPower = pyRoundN(rep.weightedAvgPower, 1);
    if (rep.weightedAvgPower && rep.avgPrimary && kind === "power") {
      rep.variabilityRatio = pyRoundN(rep.weightedAvgPower / rep.avgPrimary, 3);
    }
    const kj = workKj(session.records.time, pw.values);
    rep.workKj = kj !== null ? pyRoundN(kj, 1) : null;
    const curve = meanMax(pw.values, POWER_CURVE_WINDOWS);
    rep.powerCurve = new Map(
      [...curve.entries()].map(([w, v]) => [w, v !== null ? pyRoundN(v, 1) : null]),
    );
  }

  if (profile.distanceFromLengths && session.lengths.length > 0) {
    rep.swolf = swolf(session)[1];
  }

  rep.splits =
    profile.paceStyle === "per_km"
      ? distanceSplits(session, 1000.0, { style: profile.paceStyle })
      : [];
  rep.structure = detectStructure(session, messages, settings);

  // zones: only from settings or the file, never estimated (ADR-0008 §4)
  const [hb, hbasis] = hrZoneBounds(settings, messages);
  const hrStream = getStream(session.records, "heart_rate");
  if (hb !== null && hrStream !== undefined) {
    rep.hrZones = {
      bounds: [...hb],
      basis: hbasis,
      seconds: timeInZones(session.records.time, hrStream.values, [...hb]),
    };
  } else if (hrStream !== undefined) {
    rep.omissions.push("hr_zones: no zone bounds in settings or file");
  }
  const [pb, pbasis] = powerZoneBounds(settings, messages);
  if (pb !== null && pw !== undefined) {
    rep.powerZones = {
      bounds: [...pb],
      basis: pbasis,
      seconds: timeInZones(session.records.time, pw.values, [...pb]),
    };
  } else if (pw !== undefined) {
    rep.omissions.push("power_zones: no zone bounds in settings or file");
  }

  rep.load = workoutLoad(session, settings);
  if (rep.load === null) {
    if (pw !== undefined && (settings === null || !settings.ftpW)) {
      rep.omissions.push("load: power present but no ftp_w in settings");
    } else if (hrStream !== undefined) {
      if (settings === null || !(settings.maxHr && settings.restingHr)) {
        rep.omissions.push("load: hr present but no max_hr+resting_hr in settings");
      } else {
        const cov = hrCoverageFraction(session);
        if (cov !== null && cov < TRIMP_MIN_COVERAGE) {
          rep.omissions.push(
            `load: hr covers only ${pyFixed(cov * 100, 0)}% of the session (< ${pyFixed(TRIMP_MIN_COVERAGE * 100, 0)}%); trimp would understate load`,
          );
        }
      }
    }
  }

  for (const maybe of [
    pacingInsight(session, stream),
    hrDriftInsight(session, stream),
    coastingInsight(session),
  ]) {
    if (maybe !== null) rep.insights.push(maybe);
  }
  if (rep.structure !== null && rep.structure.repeats.length > 0) {
    rep.insights.push({
      code: "WORKOUT_STRUCTURE",
      message: `Interval structure: ${rep.structure.repeats.map((g) => g.label).join("; ")}`,
      evidence: {
        basis: rep.structure.basis,
        labels: rep.structure.repeats.map((g) => g.label),
      },
    });
  }
  return rep;
}

/** Report per session from a ParseResult (uses `.activity` and `.messages`). */
export function analyze(
  result: { activity: unknown; messages: Message[] },
  settings: AthleteSettings | null = null,
): ActivityReport {
  const activity = result.activity as { sessions: Session[] } | null;
  const messages = result.messages;
  if (activity === null) return { sessions: [] };
  return {
    sessions: activity.sessions.map((s) => analyzeSession(s, messages, settings)),
  };
}

// --- JSON serialization (the CLI's --json path) ---------------------------

/**
 * Python-int-typed report fields. Everything else numeric is a Python float, and
 * `json.dumps` writes an integral float as `119.0` where `JSON.stringify` would
 * give `119` — so the serializer needs to know which is which by key.
 */
const INT_FIELDS = new Set(["index", "count", "first_index", "step_index", "lengths"]);

function isoPlus00(t: number | null): string | null {
  // Python's datetime.isoformat() on a tz-aware UTC value: "+00:00", not "Z", and
  // fractional times carry microseconds — datetime.fromtimestamp rounds the
  // fraction to the nearest microsecond (half-even), and isoformat prints
  // ".%06d" only when nonzero.
  if (t === null) return null;
  let whole = Math.floor(t);
  let micro = pyRound((t - whole) * 1e6);
  if (micro >= 1_000_000) {
    whole += 1;
    micro = 0;
  }
  const base = fitTsToIso(whole).slice(0, -1); // strip Z
  const frac = micro > 0 ? `.${String(micro).padStart(6, "0")}` : "";
  return `${base}${frac}+00:00`;
}

/** The `to_dict()` plain tree, snake_case keys in dataclass field order. */
export function reportToPlain(report: ActivityReport): unknown {
  return { sessions: report.sessions.map(sessionPlain) };
}

function splitPlain(s: Split): Record<string, unknown> {
  return {
    index: s.index,
    start_m: s.startM,
    end_m: s.endM,
    duration_s: s.durationS,
    avg_speed_mps: s.avgSpeedMps,
    pace_s: s.paceS,
    avg_hr: s.avgHr,
    avg_power: s.avgPower,
    ascent_m: s.ascentM,
    descent_m: s.descentM,
    partial: s.partial,
  };
}

function structurePlain(st: IntervalStructure): Record<string, unknown> {
  return {
    basis: st.basis,
    intervals: st.intervals.map((iv) => ({
      index: iv.index,
      kind: iv.kind,
      start_time: isoPlus00(iv.startTime),
      end_time: isoPlus00(iv.endTime),
      duration_s: iv.durationS,
      distance_m: iv.distanceM,
      avg_primary: iv.avgPrimary,
      avg_hr: iv.avgHr,
      lengths: iv.lengths,
      step_index: iv.stepIndex,
    })),
    repeats: st.repeats.map((g) => ({
      count: g.count,
      kind: g.kind,
      mean_duration_s: g.meanDurationS,
      mean_distance_m: g.meanDistanceM,
      mean_primary: g.meanPrimary,
      mean_rest_s: g.meanRestS,
      label: g.label,
      first_index: g.firstIndex,
    })),
    note: st.note,
  };
}

function sessionPlain(rep: WorkoutReport): Record<string, unknown> {
  return {
    sport: rep.sport,
    sub_sport: rep.subSport,
    profile: rep.profile,
    primary_signal: rep.primarySignal,
    duration_s: Object.fromEntries(rep.durationS),
    distance_m: rep.distanceM,
    pace: rep.pace,
    avg_speed_kmh: rep.avgSpeedKmh,
    avg_speed_basis: rep.avgSpeedBasis,
    avg_primary: rep.avgPrimary,
    max_primary: rep.maxPrimary,
    avg_hr: rep.avgHr,
    max_hr: rep.maxHr,
    cadence: rep.cadence,
    weighted_avg_power: rep.weightedAvgPower,
    variability_ratio: rep.variabilityRatio,
    work_kj: rep.workKj,
    power_curve:
      rep.powerCurve === null
        ? null
        : Object.fromEntries([...rep.powerCurve.entries()].map(([w, v]) => [String(w), v])),
    swolf: rep.swolf,
    splits: rep.splits.map(splitPlain),
    structure: rep.structure === null ? null : structurePlain(rep.structure),
    hr_zones: rep.hrZones,
    power_zones: rep.powerZones,
    load:
      rep.load === null
        ? null
        : { value: rep.load.value, basis: rep.load.basis, components: rep.load.components },
    insights: rep.insights.map((i) => ({
      code: i.code,
      message: i.message,
      evidence: i.evidence,
    })),
    omissions: [...rep.omissions],
  };
}

/**
 * `json.dumps(tree, sort_keys=True, separators=(",", ":"), ensure_ascii=False)`,
 * with Python's float/int distinction recovered from the INT_FIELDS key set.
 */
export function dumpsReportJson(tree: unknown): string {
  return emit(tree, null);
}

function emit(v: unknown, key: string | null): string {
  if (v === null || v === undefined) return "null";
  if (typeof v === "boolean") return v ? "true" : "false";
  if (typeof v === "string") return JSON.stringify(v);
  if (typeof v === "number") {
    if (Number.isInteger(v) && !INT_FIELDS.has(key ?? "")) {
      // A Python float serializes with its .0; report ints are the named fields.
      return `${v}.0`;
    }
    return String(v);
  }
  if (Array.isArray(v)) return `[${v.map((x) => emit(x, key)).join(",")}]`;
  const entries = Object.entries(v as Record<string, unknown>).sort(([a], [b]) =>
    a < b ? -1 : a > b ? 1 : 0,
  );
  return `{${entries.map(([k, val]) => `${JSON.stringify(k)}:${emit(val, k)}`).join(",")}}`;
}
